### PR TITLE
Use matrix for gradient.

### PR DESCRIPTION
**Pull request by trivialfis**
- Use the `linalg::Matrix` for storing gradients.
- New API `XGBoosterTrainOneIter` for the custom objective. Deprecating the `XGBoosterBoostOneIter`
- Custom objective for multi-class/multi-target is now required to return the correct shape.
- Custom objective for Python can accept arrays with any strides. (row-major, column-major) In addition, CUDA input is now supported.

The new function can be extended to irregular shape of input gradient by changing the JSON array interface into a dictionary from an array.

Close https://github.com/dmlc/xgboost/issues/9089 .

### CODE REVIEW

**Comment by trivialfis** [comment]
@rongou Would you like to help review this guy when time allows?

**Comment by trivialfis** [comment]
@wbo4958 Could you please help take a look into the changes in the JVM packages?

File: R-package/R/utils.R [hunk 1]
```diff
@@ -154,7 +154,12 @@ xgb.iter.update <- function(booster_handle, dtrain, iter, obj) {
     pred <- predict(booster_handle, dtrain, outputmargin = TRUE, training = TRUE,
                     ntreelimit = 0)
     gpair <- obj(pred, dtrain)
-    .Call(XGBoosterBoostOneIter_R, booster_handle, dtrain, gpair$grad, gpair$hess)
+    n_samples <- dim(dtrain)[1]
```
Comment (trivialfis) [inline R-package/R/utils.R:157]: R package is still required to provide row-based matrix, as I don't know how to interpret the pointer returned by R.

File: R-package/src/xgboost_R.cc [hunk 2]
```diff
@@ -394,21 +387,25 @@ XGB_DLL SEXP XGBoosterUpdateOneIter_R(SEXP handle, SEXP iter, SEXP dtrain) {
...
-                                 len));
+  CHECK_EQ(length(grad), length(hess)) << "gradient and hess must have same length";
+  SEXP gdim = getAttrib(grad, R_DimSymbol);
+  auto n_samples = static_cast<std::size_t>(INTEGER(gdim)[0]);
+  auto n_targets = static_cast<std::size_t>(INTEGER(gdim)[1]);
+
+  SEXP hdim = getAttrib(hess, R_DimSymbol);
+  CHECK_EQ(INTEGER(hdim)[0], n_samples) << "mismatched size between gradient and hessian";
+  CHECK_EQ(INTEGER(hdim)[1], n_targets) << "mismatched size between gradient and hessian";
+  double const *d_grad = REAL(grad);
```
Comment (trivialfis) [inline R-package/src/xgboost_R.cc:400]: zero-copy now.

File: include/xgboost/base.h [hunk 3]
```diff
@@ -274,8 +274,8 @@ class GradientPairInt64 {
   GradientPairInt64(GradientPairInt64 const &g) = default;
   GradientPairInt64 &operator=(GradientPairInt64 const &g) = default;
 
-  XGBOOST_DEVICE [[nodiscard]] T GetQuantisedGrad() const { return grad_; }
-  XGBOOST_DEVICE [[nodiscard]] T GetQuantisedHess() const { return hess_; }
+  [[nodiscard]] XGBOOST_DEVICE T GetQuantisedGrad() const { return grad_; }
```
Comment (trivialfis) [inline include/xgboost/base.h:277]: Asked by clang tidy.

File: include/xgboost/c_api.h [hunk 4]
```diff
@@ -945,21 +943,29 @@ XGB_DLL int XGBoosterUpdateOneIter(BoosterHandle handle, int iter, DMatrixHandle
...
-XGB_DLL int XGBoosterBoostOneIter(BoosterHandle handle,
-                                  DMatrixHandle dtrain,
-                                  float *grad,
-                                  float *hess,
-                                  bst_ulong len);
+XGB_DLL int XGBoosterBoostOneIter(BoosterHandle handle, DMatrixHandle dtrain, float *grad,
+                                  float *hess, bst_ulong len);
+
+/**
+ * @brief Update a multi-target model with gradient and Hessian. This is used for training
```
Comment (trivialfis) [inline include/xgboost/c_api.h:953]:
```suggestion
 * @brief Update a model with gradient and Hessian. This is used for training
```

File: include/xgboost/gbm.h [hunk 5]
```diff
@@ -70,22 +70,24 @@ class GradientBooster : public Model, public Configurable {
...
-   * the booster may change content of gpair
+  [[nodiscard]] virtual bool ModelFitted() const = 0;
+  /**
+   * @brief perform update to the model(boosting)
+   *
+   * @param p_fmat feature matrix that provide access to features
+   * @param in_gpair address of the gradient pair statistics of the data
+   * @param prediction The output prediction cache entry that needs to be updated.
+   *                   the booster may change content of gpair
    */
```
Comment (trivialfis) [inline include/xgboost/gbm.h:89]:
```suggestion
   * @param obj Objective function.
   */
```

File: python-package/xgboost/core.py [hunk 6]
```diff
@@ -2070,19 +2072,52 @@ def boost(self, dtrain: DMatrix, grad: np.ndarray, hess: np.ndarray) -> None:
...
 
+        def is_flatten(array: NumpyOrCupy) -> bool:
+            return len(array.shape) == 1 or array.shape[1] == 1
+
+        def array_interface(array: NumpyOrCupy) -> bytes:
+            msg = (
+                "Expecting `np.ndarray` or `cupy.ndarray` for gradient and hessian."
+                f" Got: {type(array)}"
+            )
+            if not isinstance(array, np.ndarray) and not _is_cupy_array(array):
```
Comment (trivialfis) [inline python-package/xgboost/core.py:2092]: perhaps we should check for `__array_interface__` instead. But I will leave that to the future due to difficulty on testing.

File: python-package/xgboost/core.py [hunk 7]
```diff
@@ -2070,19 +2072,52 @@ def boost(self, dtrain: DMatrix, grad: np.ndarray, hess: np.ndarray) -> None:
...
+            n_samples = dtrain.num_row()
+            if array.shape[0] != n_samples and is_flatten(array):
+                warnings.warn(
+                    "Since 2.1.0, the shape of the gradient and hessian is required to"
+                    " be (n_samples, n_targets) or (n_samples, n_classes).",
+                    FutureWarning,
+                )
+                array = array.reshape(n_samples, array.size // n_samples)
+
+            if isinstance(array, np.ndarray):
```
Comment (trivialfis) [inline python-package/xgboost/core.py:2104]: We can probably consolidate the logic into a single function that can be reused for `inplace_predict`.

File: src/c_api/c_api_utils.h [hunk 8]
```diff
@@ -334,6 +349,36 @@ void MakeSparseFromPtr(PtrT const *p_indptr, I const *p_indices, T const *p_data
   Json::Dump(jindices, indices_str);
   Json::Dump(jdata, data_str);
 }
+
+template <typename G, typename H>
```
Comment (trivialfis) [inline src/c_api/c_api_utils.h:353]: Consistent use of type name `GT`, `HT`.

File: src/gbm/gbtree.cc [hunk 9]
```diff
@@ -185,16 +177,19 @@ void GPUCopyGradient(HostDeviceVector<GradientPair> const*, bst_group_t, bst_gro
 }
 #endif
 
-void CopyGradient(HostDeviceVector<GradientPair> const* in_gpair, int32_t n_threads,
-                  bst_group_t n_groups, bst_group_t group_id,
-                  HostDeviceVector<GradientPair>* out_gpair) {
-  if (in_gpair->DeviceIdx() != Context::kCpuId) {
-    GPUCopyGradient(in_gpair, n_groups, group_id, out_gpair);
+void CopyGradient(Context const* ctx, linalg::Matrix<GradientPair> const* in_gpair,
```
Comment (trivialfis) [inline src/gbm/gbtree.cc:180]: We will remove this function once we can ensure all objectives are using f-order.

**Review by rongou — APPROVED** [review]

File: plugin/example/custom_obj.cc [hunk 10]
```diff
@@ -33,38 +31,38 @@ class MyLogistic : public ObjFunction {
...
-    std::vector<GradientPair>& out_gpair_h = out_gpair->HostVector();
+  void GetGradient(const HostDeviceVector<float>& preds, MetaInfo const& info,
+                   std::int32_t /*iter*/, linalg::Matrix<GradientPair>* out_gpair) override {
+    out_gpair->Reshape(info.num_row_, 1);
+    const std::vector<float>& preds_h = preds.HostVector();
+    auto out_gpair_h = out_gpair->HostView();
     auto const labels_h = info.labels.HostView();
     for (size_t i = 0; i < preds_h.size(); ++i) {
-      bst_float w = info.GetWeight(i);
+      float w = info.GetWeight(i);
```
[the remark this answers is not in the document]
Comment (trivialfis) [reply plugin/example/custom_obj.cc:43]: I don't have a preference, just trying to clean up some of these old types.
